File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        TreeMap<Integer,Integer> mp=new TreeMap<>();
        TreeMap<Integer,Integer> mp1=new TreeMap<>();
        int l1=s1.length();
        int l2=s2.length();
        if(l1>l2)return false;
        for(int i=0;i<l1;i++)
        {
            int t1=s1.charAt(i)-'A';
            int t2=s2.charAt(i)-'A';
            mp.put(t1,mp.getOrDefault(t1,0)+1);
            mp1.put(t2,mp1.getOrDefault(t2,0)+1);
        }
        if(mp.equals(mp1))return true;
        for(int i=l1;i<l2;i++)
        {
            int r=s2.charAt(i)-'A';
            int q=s2.charAt(i-l1)-'A';
            mp1.put(q,mp1.get(q)-1);
            if(mp1.get(q)==0)
            mp1.remove(s2.charAt(i-l1)-'A');
            mp1.put(r,mp1.getOrDefault(r,0)+1);
            if(mp.equals(mp1))
            return true;
        }
        return false;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:

    bool checkInclusion(string s1, string s2) {
        unordered_map<char,int> mp;
        unordered_map<char,int> mp1;
        for(char c:s1)
        mp[c]++;
        int n=s1.size();
        int l=0;
        int r=0;
        int n1=s2.size();
        if(n > n1) return false;
        while(r-l<n)
        {mp1[s2[r]]++;r++;}
        if(mp==mp1)
            return true;
        while(r < n1) {
            mp1[s2[l]]--;
            if(mp1[s2[l]] == 0) mp1.erase(s2[l]);
            l++;
            mp1[s2[r]]++;
            r++;
            if(mp == mp1) return true;
        }
        return false;
    }
};